arbnode/batch_poster.go

```diff
@@ -496,7 +496,21 @@ func (b *BatchPoster) addEspressoBlockMerkleProof(
 	ctx context.Context,
 	msg *arbostypes.MessageWithMetadata,
 ) error {
+
 	if arbos.IsEspressoMsg(msg.Message) {
+		arbOSConfig, err := b.arbOSVersionGetter.GetArbOSConfigAtHeight(0)
+		if err != nil {
+			return fmt.Errorf("Failed call to GetArbOSConfigAtHeight: %w", err)
+		}
+		if arbOSConfig == nil {
+			return fmt.Errorf("Cannot use a nil ArbOSConfig")
+		}
+		if !arbOSConfig.ArbitrumChainParams.EnableEspresso {
```

Comment:
Don't want to derail this but I just started wondering if things were simpler if instead of having an `EnableEspresso` flag we had something like a `EnableEspressoAtChildBlock` field instead.

I think this might allow an operator to 

1. upgrade all the software 
2. schedule the switch to Espresso to happen via a contract upgrade some time in the future 
3. update the configuration of all components with `EnableEspressoAtChildBlock` to match what we set during the contract upgrade

When L2 block reaches `EnableEspressoAtChildBlock` the Espresso logic turns on everywhere.

Then I think the components wouldn't need to be able to read live `ArbitrumChainParams`.

@zacshowa @Sneh1999 @ImJeremyHe what do you think?

Reply:
@sveitser  would this mean that all components will need to have access to L2 block number which they will have to fetch live as well?

I see how this could give sometime to the operator to upgrade all the software and make any necessary changes before the switch happens which will not be the case if switch happens the moment we change the chain config.

Reply:
I think this is an interesting proposal, but I have a few questions.
Fundamentally, if we are using `SetChainConfig()` on the `ArbOwner` precompile, we will need to have some component reading from the ArbOS config after the transaction is sent to update the `ArbitrumChainParams`. Currently the various components fetch this and read from it based on the current block to decide this information on the `EnableEspresso` boolean.
I think it could enable more ergonomic control from the node operator perspective to be able to set a specific block height in which the changes are enabled..
However, An implementation isn't clear to me. Specifically, how can we read from the ArbOS chain config better than checking it every block. My first thought was to read it on startup. However, reading the value once on node startup ans saving it to a local config has problems. It would introduce a required on nodes after calls `setChainConfig()` on `ArbOwner` and make this an ineffective runtime switch.